### PR TITLE
Implement challenge.tar.gz unzip pipeline

### PR DESCRIPTION
**Pull request by Altoiids**

### CODE REVIEW

File: lib/utils/docker.go [hunk 1]
```diff
@@ -50,6 +55,10 @@ func BuildDockerImage(_ChallengeName string, _DockerfilePath string) {
 		return
 	}
 
+	log.Println(_ChallengeName)
```
Comment (ashpect) [inline lib/utils/docker.go:58]: Remove logging

File: services/flaghandlerservice/helper.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	//regex to find challenge name since localFilePath[12:22] is hardcoded
+	// regexPattern := `\/([^\/]+)\.tar\.gz$`
+	// regex := regexp.MustCompile(regexPattern)
+	// matches := regex.FindStringSubmatch(localFilePath)
+	// filename := matches[1]
+
+	// Get pods from different namespaces
+	
+		// Copy file into pod
+		if err := utils.CopyIntoPod(podName, g.TeamVmConfig.ContainerName, pathInPod, localFilePath, namespace); err != nil {
```
Comment (ashpect) [inline services/flaghandlerservice/helper.go:29]: This will throw error obv, no func copyinpod remaining, change required name of func her for add the func.

File: services/challengedeployerservice/helper.go [hunk 3]
```diff
@@ -117,3 +113,28 @@ func createFolder(challengeName string) (message int, challengePath string) {
 	//Successfully created directory
 	return 0, challengePath
 }
+
+func copyChallengeCheckerIntoKissaki(dirPath string, challengeName string) error {
+	srcFilePath := dirPath + "/" + challengeName + "_challenge_checker"
+	pathInPod := "/opt/kissaki/challenge-data"
+
+	log.Println("Testing" + srcFilePath + "....and..." + pathInPod)
```
Comment (ashpect) [inline services/challengedeployerservice/helper.go:121]: remove

File: services/challengedeployerservice/helper.go [hunk 4]
```diff
@@ -117,3 +113,28 @@ func createFolder(challengeName string) (message int, challengePath string) {
...
+		log.Println(err)
+		return err
+	}
+	return nil
+}
+
+func copyFlagDataIntoKashira(dirPath string, challengeName string) error {
+	srcFilePath := dirPath + "/" + "flag_data"
+	pathInPod := "/opt/kashira/flag-data"
+	log.Println("Testing" + srcFilePath + "....and..." + pathInPod)
```
Comment (ashpect) [inline services/challengedeployerservice/helper.go:133]: either write proper logging like the scrfile is : <filepth> or remove undirected printstateents

File: lib/utils/docker.go [hunk 5]
```diff
@@ -38,6 +38,11 @@ func dockerLogin(username string, password string) {
 	log.Println("Logged into Harbor successfully")
 }
 
+func CheckDockerfile(_DockerfilePath string) bool {
```
Comment (Bisht13) [inline lib/utils/docker.go:41]: Not needed, stat the files wherever it is required directly

File: lib/utils/kube.go [hunk 6]
```diff
@@ -428,3 +352,90 @@ func GetNodes(clientset *kubernetes.Clientset) ([]corev1.Node, error) {
...
+	namespace := "katana"
+	if len(ns) > 0 {
+		namespace = ns[0]
+	}
+
+	reader, writer := io.Pipe()
+
+	var cmdArr []string
+
+	cmdArr = []string{"tar", "-zxf", "-"}
```
Comment (Bisht13) [inline lib/utils/kube.go:376]: Don't tar using cmd

File: services/challengedeployerservice/controller.go [hunk 7]
```diff
@@ -18,11 +17,75 @@ import (
 	"github.com/sdslabs/katana/lib/utils"
 	"github.com/sdslabs/katana/types"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+    _ "unsafe"
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:20]: What is this for?

File: services/challengedeployerservice/controller.go [hunk 8]
```diff
@@ -18,11 +17,75 @@ import (
...
 )
+//for get request
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:22]: Remove comment

File: services/challengedeployerservice/controller.go [hunk 9]
```diff
@@ -18,11 +17,75 @@ import (
...
+func Deploy(c *fiber.Ctx) error {
+	patch := false
+	replicas := int32(1)
+	log.Println("Starting")
+	//read folder challenge by os
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:27]: Keep consistency in your comments, some are starting with lower case while others are using upper case, look at the code base and find out how it is done

File: services/challengedeployerservice/controller.go [hunk 10]
```diff
@@ -18,11 +17,75 @@ import (
...
+		if fileInfo.IsDir() {
+			//Get the challenger name
+			folderName := fileInfo.Name()
+			log.Println("Folder name is : " + folderName)
+			//update challenge path to be absolute path
+			challengePath, _ := os.Getwd()
+			challengePath = challengePath + "/challenges/" + folderName
+			log.Println("Challenge path is : " + challengePath)
+			//check if the folder has a Dockerfile
+			if utils.CheckDockerfile(challengePath + "/" + folderName) {
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:56]: Why is this check required?

File: services/challengedeployerservice/controller.go [hunk 11]
```diff
@@ -18,11 +17,75 @@ import (
...
+				log.Println("Dockerfile not found in the " + folderName + " challenge folder. Please follow proper format.")
+			}
+			copyChallengeIntoTsuka(challengePath,folderName)
+			copyFlagDataIntoKashira(challengePath,folderName)
+			copyChallengeCheckerIntoKissaki(challengePath,folderName)
+		}
+	}
+	return c.JSON(res)
+}
+
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:84]: Remove extra line

File: services/challengedeployerservice/controller.go [hunk 12]
```diff
@@ -87,14 +149,16 @@ func DeployChallenge(c *fiber.Ctx) error {
 				}
 			}
 
-			//Copy challenge in pods and etc.
-			copyChallengeIntoTsuka(challengePath, folderName, challengeType)
+			copyChallengeIntoTsuka(challengePath,folderName)
+			copyFlagDataIntoKashira(challengePath,folderName)
+			copyChallengeCheckerIntoKissaki(challengePath,folderName)
+			
+
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:156]: Remove extra lines

File: services/flaghandlerservice/helper.go [hunk 13]
```diff
@@ -0,0 +1 @@
+package flaghandlerservice
```
Comment (Bisht13) [inline services/flaghandlerservice/helper.go:1]: Why this file?

File: challenge_template/knock/flag_data/knock/knock_flag_getter.txt [hunk 14]
```diff
@@ -0,0 +1 @@
+knock flag getter
```
Comment (Bisht13) [inline challenge_template/knock/flag_data/knock/knock_flag_getter.txt:1]: File name should contain - not _

File: challenge_template/knock/flag_data/knock/knock_flag_setter.txt [hunk 15]
```diff
@@ -0,0 +1 @@
+knock flag setter
```
Comment (Bisht13) [inline challenge_template/knock/flag_data/knock/knock_flag_setter.txt:1]: File name should contain - not _

File: challenges/.gitkeep
Comment (Bisht13) [inline challenges/.gitkeep:1]: Why remove this file?

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: lib/utils/docker.go (hunk 5 above)
Comment (Altoiids) [reply lib/utils/docker.go:41]: done

File: lib/utils/kube.go (hunk 6 above)
Comment (Altoiids) [reply lib/utils/kube.go]: changed the implementation as discussed

File: services/challengedeployerservice/controller.go (hunk 7 above)
Comment (Altoiids) [reply services/challengedeployerservice/controller.go]: Forgot to remove

File: services/challengedeployerservice/controller.go (hunk 8 above)
Comment (Altoiids) [reply services/challengedeployerservice/controller.go]: Done

File: services/challengedeployerservice/controller.go (hunk 9 above)
Comment (Altoiids) [reply services/challengedeployerservice/controller.go]: Done

File: services/challengedeployerservice/controller.go (hunk 12 above)
Comment (Altoiids) [reply services/challengedeployerservice/controller.go]: Done

File: services/flaghandlerservice/helper.go (hunk 13 above)
Comment (Altoiids) [reply services/flaghandlerservice/helper.go:1]: It was already present in this form.

File: challenge_template/knock/flag_data/knock/knock_flag_getter.txt (hunk 14 above)
Comment (Altoiids) [reply challenge_template/knock/flag_data/knock/knock_flag_getter.txt]: corrected

File: challenge_template/knock/flag_data/knock/knock_flag_setter.txt (hunk 15 above)
Comment (Altoiids) [reply challenge_template/knock/flag_data/knock/knock_flag_setter.txt]: Corrected

File: challenges/.gitkeep
Comment (Altoiids) [reply challenges/.gitkeep:1]: Because it was not required. I have pushed challenge_template instead for sample challenges.

File: lib/utils/kube.go [hunk 16]
```diff
@@ -22,6 +24,7 @@ import (
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/tools/clientcmd"
 	"k8s.io/client-go/tools/remotecommand"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
```
[the remark this answers is not in the document]
Comment (Altoiids) [reply lib/utils/kube.go:25]: Done

File: services/flaghandlerservice/helper.go (hunk 2 above)
Comment (Altoiids) [reply services/flaghandlerservice/helper.go]: I have removed it but forgot to push this file. Corrected now.

File: services/challengedeployerservice/helper.go (hunk 4 above)
Comment (Altoiids) [reply services/challengedeployerservice/helper.go]: done

File: challenge_template/knock/knock.tar.gz
Comment (Bisht13) [inline challenge_template/knock/knock.tar.gz:1]: why this file?

File: challenges/.gitkeep
Comment (Bisht13) [inline challenges/.gitkeep:1]: keep this file

File: kubernetes/manifests/kashira.yml [hunk 17]
```diff
@@ -20,12 +28,20 @@ spec:
       serviceAccountName: cluster-admin
       containers:
         - name: kashira
-          image: harbor.katana.local/katana/kashira
+          image: ashpect/kashira
```
Comment (Bisht13) [inline kubernetes/manifests/kashira.yml:31]: fix this

File: kubernetes/manifests/kashira.yml [hunk 18]
```diff
@@ -20,12 +28,20 @@ spec:
...
           ports:
             - containerPort: 80
           volumeMounts:
             - name: kashira-data
               mountPath: /opt/kashira/flag-data
+
```
Comment (Bisht13) [inline kubernetes/manifests/kashira.yml:37]: extra line not needed

File: kubernetes/manifests/kashira.yml [hunk 19]
```diff
@@ -20,12 +28,20 @@ spec:
...
+          env:
+            - name: DAEMON_PORT
+              valueFrom:
+                configMapKeyRef:
+                  name: kashira-config
+                  key: daemon_port
+
```
Comment (Bisht13) [inline kubernetes/manifests/kashira.yml:44]: extra line not needed

File: kubernetes/manifests/kashira.yml [hunk 20]
```diff
@@ -1,3 +1,11 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: kashira-config
+  namespace: katana
+data:
+  daemon_port: "{{.DaemonPort}}"
```
Comment (Bisht13) [inline kubernetes/manifests/kashira.yml:7]: Why is this required?

File: kubernetes/manifests/kissaki.yml [hunk 21]
```diff
@@ -20,12 +28,19 @@ spec:
       serviceAccountName: cluster-admin
       containers:
         - name: kissaki
-          image: harbor.katana.local/katana/kissaki
+          image: ashpect/kissaki
```
Comment (Bisht13) [inline kubernetes/manifests/kissaki.yml:31]: Same issues

File: lib/utils/crypto.go [hunk 22]
```diff
@@ -64,6 +63,154 @@ func GenerateCerts(domain string, basePath string) error {
 	return nil
 }
 
+// [DO NOT REMOVE....WIP....]
```
Comment (Bisht13) [inline lib/utils/crypto.go:66]: How can I merge in `develop` then?

File: lib/utils/docker.go (hunk 5 above)
Comment (Bisht13) [reply lib/utils/docker.go:41]: Still present

File: lib/utils/kube.go [hunk 23]
```diff
@@ -428,3 +352,85 @@ func GetNodes(clientset *kubernetes.Clientset) ([]corev1.Node, error) {
 
 	return nodes.Items, nil
 }
+
+func CopyTarIntoPod(podName string, containerName string, pathInPod string, localFilePath string, ns ...string) error {
```
Comment (Bisht13) [inline lib/utils/kube.go:356]: Isn't this the same function? why change the name then?

File: lib/utils/os.go [hunk 24]
```diff
@@ -77,4 +77,4 @@ func GetKatanaRootPath() (string, error) {
 		return "", nil
 	}
 	return katanaDir, nil
-}
+}
```
Comment (Bisht13) [inline lib/utils/os.go:80]: What is this change?

File: services/challengedeployerservice/controller.go [hunk 25]
```diff
@@ -18,10 +12,81 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"log"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
 )
 
-func DeployChallenge(c *fiber.Ctx) error {
+func Deploy(c *fiber.Ctx) error {
```
Comment (Bisht13) [inline services/challengedeployerservice/controller.go:22]: Why is this function changed?

File: services/flaghandlerservice/helper.go (hunk 13 above)
Comment (Bisht13) [reply services/flaghandlerservice/helper.go:1]: You are adding it via this PR

**Review by Bisht13 — CHANGES_REQUESTED** [review]

File: kubernetes/manifests/kashira.yml (hunk 19 above)
Comment (Altoiids) [reply kubernetes/manifests/kashira.yml:44]: removed

File: services/challengedeployerservice/controller.go (hunk 25 above)
Comment (Altoiids) [reply services/challengedeployerservice/controller.go:22]: Deploy is for get request.
DeployChallenge is for post request. 
Was testing something. Now restored

**Comment by Altoiids** [comment]
removed

File: challenge_template/knock/knock.tar.gz
Comment (Altoiids) [reply challenge_template/knock/knock.tar.gz:1]: Already present in the sample challenge provided by ashpect

File: lib/utils/kube.go (hunk 23 above)
Comment (Altoiids) [reply lib/utils/kube.go:356]: changed to original

File: kubernetes/manifests/kashira.yml (hunk 20 above)
Comment (Altoiids) [reply kubernetes/manifests/kashira.yml:7]: Because percy said not to hardcode so I have set the env in yml file
